### DISCUSSION TITLE
Add custom constraints in MessageIX

### DISCUSSION THREAD

**Discussion by Alaa-Eltablawy** [iiasa/message_ix#840] [MESSAGE modeling]
Hello!

Is it possible to add custom constraints based on existing variables in MessageIX Python API?

Thanks!

**Comment by khaeru** [comment]
@Alaa-Eltablawy, do you mean “custom” constraints?

If you do, please edit your question & title.
If not, please provide a link or explanation of what “costume constraints” are.

**Reply by khaeru** [reply]
Okay, it seems "custom" was indeed intended.

The answer is **yes, in some cases**. See the sections of the documentation about:
- [“Generic relations (linear constraints)”](https://docs.messageix.org/en/latest/model/MESSAGE/model_core.html#section-of-generic-relations-linear-constraints) —this describes the GAMS variables and equations for these constraints, or ‘relations’.
- [“Parameteters of the *Relations* section”](https://docs.messageix.org/en/latest/model/MESSAGE/parameter_def.html#parameters-of-the-relations-section) —this describes the particular parameters in which data should be entered.

In general, per the first link, it is possible to construct a constraint on nearly any linear combination of `ACT`, `CAP`, and `CAP_NEW` values.

Since these are the ‘core’ decision variables for MESSAGE, the values of most other variables are closely related, and it is possible to effectively constrain them by careful construction of values for the parameters `relation_activity`, `relation_new_capacity`, and `relation_total_capacity`.
